include/heyoka/mascon.hpp

```diff
@@ -0,0 +1,212 @@
+// with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#ifndef HEYOKA_MASCON_HPP
+#define HEYOKA_MASCON_HPP
+
+#include <heyoka/detail/igor.hpp>
+#include <heyoka/expression.hpp>
+#include <heyoka/kw.hpp>
+#include <heyoka/math_functions.hpp>
+#include <heyoka/square.hpp>
+#include <vector>
+
+namespace heyoka
+{
+
+// mascon_points -> [N,3] array containing the positions of the masses (units L)
+// mascon_masses -> [N] array containing the position of the masses (units M)
+// pd, qd, rd -> angular velocity of the asteroid in the frame used for the mascon model (units rad/T)
+//
+// GConst kwarg -> Cavendish constant (units L^3/T^2/M)
+// Note, units must be consistent. Choosing L and M is done via the mascon model, T is derived by the value of G. The
+// angular velocity must be consequent (equivalently one can choose the units for w and induce them on the value of G).
+template <typename... KwArgs>
+inline std::vector<std::pair<expression, expression>> make_mascon_system(KwArgs &&... kw_args)
+{
+    // 1 - Check input consistency (TODO)
+    // 2 - We parse the unnamed arguments
+    igor::parser p{kw_args...};
+    if constexpr (p.has_unnamed_arguments()) {
+        static_assert(detail::always_false_v<KwArgs...>,
+                      "The variadic arguments in the construction of the mascon system contain "
+                      "unnamed arguments.");
+    } else {
+        // G constant (defaults to 1).
+        auto Gconst = [&p]() {
+            if constexpr (p.has(kw::Gconst)) {
+                return expression{number{std::forward<decltype(p(kw::Gconst))>(p(kw::Gconst))}};
+            } else {
+                return expression{number{1.}};
+            }
+        }();
+
+        // mascon_points (no default)
+        std::vector<std::vector<number>> mascon_points;
+        if constexpr (p.has(kw::mascon_points)) {
+            for (const auto &point : p(kw::mascon_points)) {
+                mascon_points.emplace_back(std::vector<number>{number{point[0]}, number{point[1]}, number{point[2]}});
+            }
+        } else {
+            throw std::invalid_argument("mascon_points is missing from the kwarg list!");
+        };
+
+        // mascon_masses (no default)
+        std::vector<number> mascon_masses;
+        if constexpr (p.has(kw::mascon_masses)) {
+            for (const auto &mass : p(kw::mascon_masses)) {
+                mascon_masses.emplace_back(mass);
+            }
+        } else {
+            throw std::invalid_argument("mascon_masses is missing from the kwarg list!");
+        };
+
+        // omega (no default)
+        number pn(0.), qn(0.), rn(0.);
+        if constexpr (p.has(kw::omega)) {
+            pn = number{p(kw::omega)[0]};
+            qn = number{p(kw::omega)[1]};
+            rn = number{p(kw::omega)[2]};
+        } else {
+            throw std::invalid_argument("omega is missing from the kwarg list!");
+        };
+
```

Comment:
After you are finished parsing the input arguments with igor, I'd suggest you defer the rest of the construction to a detail compiled function in the .cpp. The parsing has to be done in a template and thus it has to be in a header, but there's no reason to keep everything in the template, it just slows compilation down.